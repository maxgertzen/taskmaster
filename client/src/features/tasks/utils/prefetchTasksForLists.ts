import { QueryClient } from '@tanstack/react-query';

import { QUERY_KEYS } from '@/shared/api/query-keys';
import { STALE_TIME } from '@/shared/constants/staleTime';

import { fetchTasks } from '../api';

export const prefetchTasksForLists = (
  listIds: string[],
  queryClient: QueryClient,
  token: string | null
) => {
  if (!token || !listIds.length) return;

  listIds.forEach((listId) => {
    queryClient.prefetchQuery({
      queryKey: QUERY_KEYS.tasks({ listId }),
      queryFn: fetchTasks(token),
      staleTime: STALE_TIME,
    });
  });
};
