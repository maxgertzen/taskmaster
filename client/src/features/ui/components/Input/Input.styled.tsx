import styled from '@emotion/styled';

export const StyledInputContainer = styled.div<{ isSearch: boolean }>(
  ({ theme, isSearch }) => ({
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
    position: 'relative',
    flex: 14,
    flexShrink: 0,
    flexBasis: '50%',
    gap: theme.spacing(1),
    ...(isSearch ? { marginRight: theme.spacing(1) } : {}),
    svg: {
      cursor: 'pointer',
    },
    [`@media (max-width: ${theme.breakpoints.sm})`]: {
      flex: 1,
      gap: theme.spacing(0.5),
      justifyContent: 'flex-end',
      marginRight: 0,
      alignItems: 'end',
    },
  })
);

export const StyledInput = styled.input<{ isSearch: boolean }>(
  ({ theme, isSearch }) => ({
    width: '100%',
    padding: theme.spacing(isSearch ? 1 : 2),
    border: theme.borders.main(),
    borderRadius: theme.borders.radius,
    fontSize: theme.typography.body.fontSize,
    fontFamily: theme.typography.fontFamily,
    backgroundColor: theme.colors.surface,
    color: theme.colors.text,
  })
);
