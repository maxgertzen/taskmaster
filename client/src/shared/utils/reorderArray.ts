export const reorderArray = <T>(
  array: T[],
  fromIndex: number,
  toIndex: number
): T[] => {
  const reordered = [...array];
  const [movedItem] = reordered.splice(fromIndex, 1);
  reordered.splice(toIndex, 0, movedItem);
  return reordered;
};
