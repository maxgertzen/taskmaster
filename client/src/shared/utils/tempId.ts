export const generateTempId = (prefix: string = 'temp') =>
  `${prefix}-${Date.now()}`;
