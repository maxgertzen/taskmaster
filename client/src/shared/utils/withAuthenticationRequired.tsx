import { withAuthenticationRequired as realWithAuthenticationRequired } from '@auth0/auth0-react';

import { IS_AUTH0_DISABLED } from '@/constants';
import { mockWithAuthenticationRequired } from '@/mocks';

const isMock = IS_AUTH0_DISABLED === 'true';

export const withAuthenticationRequired = isMock
  ? mockWithAuthenticationRequired
  : realWithAuthenticationRequired;
