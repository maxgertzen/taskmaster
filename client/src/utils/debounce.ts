export function debounce<T extends (...args: Parameters<T>) => ReturnType<T>>(
  func: T,
  delay: number
): ((...args: Parameters<T>) => void) & { cancel: () => void } {
  let timeoutId: ReturnType<typeof setTimeout> | null = null;

  const debouncedFunction = (...args: Parameters<T>) => {
    if (timeoutId) clearTimeout(timeoutId);
    timeoutId = setTimeout(() => func(...args), delay);
  };

  debouncedFunction.cancel = () => {
    if (timeoutId) clearTimeout(timeoutId);
  };

  return debouncedFunction;
}
