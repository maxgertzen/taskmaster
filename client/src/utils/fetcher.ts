export type FetcherConfig<K> = {
  url: string;
  method?: 'GET' | 'POST' | 'PUT' | 'DELETE';
  body?: K;
  timeout?: number;
  headers?: HeadersInit;
  urlSearchParams?: string;
};

export const fetcher = async <T, K = undefined>(
  config: FetcherConfig<K>
): Promise<T> => {
  try {
    const {
      method = 'GET',
      body = {},
      headers = {},
      timeout = 5000,
      urlSearchParams,
    } = config;
    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), timeout);

    const options: RequestInit = {
      method,
      headers: {
        'Content-Type': 'application/json',
        ...headers,
      },
      ...(method !== 'GET' && { body: JSON.stringify(body) }),
      signal: controller.signal,
    };

    if (urlSearchParams) {
      config.url += `?${urlSearchParams}`;
    }

    const response = await fetch(config.url, options);
    clearTimeout(timeoutId);

    if (!response.ok) {
      throw new Error(`Error: ${response.status} ${response.statusText}`);
    }

    return await response.json();
  } catch (error) {
    console.error('Failed to fetch data:', error);
    throw error;
  }
};
