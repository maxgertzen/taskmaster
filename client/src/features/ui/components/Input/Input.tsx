import { useEffect, useMemo, useState } from 'react';

import { debounce } from '@/shared/utils/debounce';

import { HighlightedArea, SpriteIcon } from '..';

import { StyledInputContainer, StyledInput } from './Input.styled';

interface InputProps {
  placeholder: string;
  onSubmit: (text: string) => void;
  onReset?: () => void;
  isSearch?: boolean;
  value?: string;
  withToggle?: boolean;
  highlightId?: string;
  'data-testid'?: string;
}

export const Input: React.FC<InputProps> = ({
  onSubmit,
  onReset,
  placeholder,
  highlightId,
  value = '',
  withToggle = false,
  isSearch = false,
  'data-testid': dataTestId,
}) => {
  const [showInput, setShowInput] = useState<boolean>(!withToggle || !!value);
  const [text, setText] = useState<string>(value);

  const debouncedOnSubmit = useMemo(
    () =>
      debounce((text: string) => {
        onSubmit(text);
      }, 300),
    [onSubmit]
  );

  const handleOnSubmit = () => {
    if (!text) return;

    if (isSearch) {
      debouncedOnSubmit(text);
    } else {
      onSubmit(text);
      setText('');
    }

    if (onReset) {
      onReset();
    }
  };

  const toggleInput = () => {
    if (showInput && text) {
      handleOnSubmit();
    }
    setShowInput((prev) => !prev);
  };

  const handleKeyDown = ({ key }: React.KeyboardEvent<HTMLInputElement>) => {
    if (key === 'Enter' && text) {
      handleOnSubmit();
    }
  };

  const handleChange = ({
    target: { value },
  }: React.ChangeEvent<HTMLInputElement>) => {
    setText(value);

    if (isSearch) {
      debouncedOnSubmit(value);
    }
  };

  useEffect(() => {
    return () => {
      if (isSearch) {
        debouncedOnSubmit.cancel();
      }
    };
  }, [debouncedOnSubmit, isSearch]);

  useEffect(() => {
    if (isSearch && value) {
      setText(value);
    }
  }, [isSearch, value]);

  return (
    <StyledInputContainer data-testid={dataTestId} isSearch={isSearch}>
      {showInput && (
        <StyledInput
          data-testid='input'
          type='text'
          value={text}
          onChange={handleChange}
          onKeyDown={handleKeyDown}
          placeholder={placeholder}
          isSearch={isSearch}
        />
      )}
      <HighlightedArea id={highlightId ?? ''}>
        <SpriteIcon
          data-testid={isSearch ? 'search-icon' : 'add-icon'}
          name={isSearch ? 'magnifying' : 'plus'}
          size={4}
          onClick={withToggle ? toggleInput : handleOnSubmit}
        />
      </HighlightedArea>
    </StyledInputContainer>
  );
};
