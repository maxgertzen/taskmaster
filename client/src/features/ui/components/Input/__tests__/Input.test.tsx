import { fireEvent } from '@testing-library/dom';
import { act } from '@testing-library/react';
import { describe, it, expect, vi, beforeEach, afterEach } from 'vitest';

import { render, screen } from '@/tests/utils';

import { Input } from '../Input';

describe('Input', () => {
  const mockOnSubmit = vi.fn();
  const mockOnReset = vi.fn();

  beforeEach(() => {
    vi.clearAllMocks();
    vi.useFakeTimers();
  });

  afterEach(() => {
    vi.useRealTimers();
  });

  describe.each([
    ['regular', false],
    ['search', true],
  ])('%s input type', (_, isSearch) => {
    it.each([
      ['light', undefined],
      ['dark', 'dark'],
    ])('renders correctly in %s theme', (_, theme) => {
      const { container } = render(
        <Input
          placeholder='Enter text'
          onSubmit={mockOnSubmit}
          isSearch={isSearch}
        />,
        { theme: theme as 'light' | 'dark' }
      );
      expect(container).toMatchSnapshot();
    });
  });

  describe('Rendering States', () => {
    it('renders with toggle hidden initially when withToggle=true', () => {
      render(
        <Input
          placeholder='Toggle input'
          onSubmit={mockOnSubmit}
          withToggle={true}
        />
      );

      expect(screen.queryByTestId('input')).not.toBeInTheDocument();
      expect(screen.getByTestId('add-icon')).toBeInTheDocument();
    });

    it('shows input with initial value despite withToggle=true', () => {
      render(
        <Input
          placeholder='Toggle input'
          onSubmit={mockOnSubmit}
          withToggle={true}
          value='initial value'
        />
      );

      expect(screen.getByTestId('input')).toBeInTheDocument();
      expect(screen.getByTestId('add-icon')).toBeInTheDocument();
    });

    it('renders correct icon for search variant', () => {
      render(
        <Input placeholder='Search' onSubmit={mockOnSubmit} isSearch={true} />
      );

      expect(screen.getByTestId('search-icon')).toBeInTheDocument();
    });
  });

  describe('User Interactions', () => {
    it('handles text input changes', async () => {
      render(<Input placeholder='Enter text' onSubmit={mockOnSubmit} />);

      const input = screen.getByTestId('input');
      await act(async () => {
        fireEvent.change(input, { target: { value: 'test input' } });
      });

      expect(input).toHaveValue('test input');
    });

    it('submits on Enter key press', async () => {
      render(<Input placeholder='Enter text' onSubmit={mockOnSubmit} />);

      const input = screen.getByTestId('input');
      await act(async () => {
        fireEvent.change(input, { target: { value: 'test input' } });
        fireEvent.keyDown(input, { key: 'Enter' });
      });

      expect(mockOnSubmit).toHaveBeenCalledWith('test input');
    });

    it('toggles input visibility with icon when withToggle=true', async () => {
      render(
        <Input
          placeholder='Toggle input'
          onSubmit={mockOnSubmit}
          withToggle={true}
        />
      );

      const icon = screen.getByTestId('add-icon');

      await act(async () => {
        fireEvent.click(icon);
      });
      expect(screen.getByTestId('input')).toBeInTheDocument();

      await act(async () => {
        fireEvent.click(icon);
      });
      expect(screen.queryByTestId('input')).not.toBeInTheDocument();
    });

    it('calls onReset when provided', async () => {
      render(
        <Input
          placeholder='Test input'
          onSubmit={mockOnSubmit}
          onReset={mockOnReset}
        />
      );

      const input = screen.getByTestId('input');
      await act(async () => {
        fireEvent.change(input, { target: { value: 'test' } });
        fireEvent.keyDown(input, { key: 'Enter' });
      });

      expect(mockOnReset).toHaveBeenCalled();
    });
  });

  describe('Search Behavior', () => {
    it('debounces search input', async () => {
      render(
        <Input placeholder='Search' onSubmit={mockOnSubmit} isSearch={true} />
      );

      const input = screen.getByTestId('input');

      await act(async () => {
        fireEvent.change(input, { target: { value: 'search term' } });
      });

      expect(mockOnSubmit).not.toHaveBeenCalled();

      await act(async () => {
        vi.advanceTimersByTime(300);
      });

      expect(mockOnSubmit).toHaveBeenCalledWith('search term');
    });
  });

  describe('Props Updates', () => {
    it('updates input value when value prop changes', async () => {
      const { rerender } = render(
        <Input
          placeholder='Test input'
          onSubmit={mockOnSubmit}
          value='initial'
          isSearch={true}
        />
      );

      expect(screen.getByTestId('input')).toHaveValue('initial');

      rerender(
        <Input
          placeholder='Test input'
          onSubmit={mockOnSubmit}
          value='updated'
          isSearch={true}
        />
      );

      expect(screen.getByTestId('input')).toHaveValue('updated');
    });
  });
});
