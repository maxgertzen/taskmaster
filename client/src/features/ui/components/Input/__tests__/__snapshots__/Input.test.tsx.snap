// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Input > regular input type > renders correctly in dark theme 1`] = `
<div>
  <div
    class="css-1hqlkg7"
  >
    <input
      class="css-1jaynes"
      data-testid="input"
      placeholder="Enter text"
      type="text"
      value=""
    />
    <div
      class="css-1rqeb1j"
      data-spotlight=""
    >
      <div
        class="css-y8t6if"
        data-testid="add-icon"
        role="button"
        size="4"
      >
        <svg
          data-testid="plus-icon"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Input > regular input type > renders correctly in light theme 1`] = `
<div>
  <div
    class="css-1hqlkg7"
  >
    <input
      class="css-wy3ceo"
      data-testid="input"
      placeholder="Enter text"
      type="text"
      value=""
    />
    <div
      class="css-1rqeb1j"
      data-spotlight=""
    >
      <div
        class="css-y8t6if"
        data-testid="add-icon"
        role="button"
        size="4"
      >
        <svg
          data-testid="plus-icon"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Input > search input type > renders correctly in dark theme 1`] = `
<div>
  <div
    class="css-1egpccx"
  >
    <input
      class="css-1q2qwsy"
      data-testid="input"
      placeholder="Enter text"
      type="text"
      value=""
    />
    <div
      class="css-1rqeb1j"
      data-spotlight=""
    >
      <div
        class="css-y8t6if"
        data-testid="search-icon"
        role="button"
        size="4"
      >
        <svg
          data-testid="magnifying-icon"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Input > search input type > renders correctly in light theme 1`] = `
<div>
  <div
    class="css-1egpccx"
  >
    <input
      class="css-10md7mk"
      data-testid="input"
      placeholder="Enter text"
      type="text"
      value=""
    />
    <div
      class="css-1rqeb1j"
      data-spotlight=""
    >
      <div
        class="css-y8t6if"
        data-testid="search-icon"
        role="button"
        size="4"
      >
        <svg
          data-testid="magnifying-icon"
        />
      </div>
    </div>
  </div>
</div>
`;
